### PR TITLE
Support unused inputs reporting for Kernel/DDK

### PR DESCRIPTION
**Pull request by jakemac53**
Uses the new `build` api to report unused kernel dependencies.

The kernel and ddk workers support a flag to write the used dependencies to a file on disk, so we take the difference between all the transitive inputs and this list to compute the unused kernel inputs.

There are a couple defensive mechanisms built into this as well:

- We never report unused deps if there is a failure.
- We never report unused deps if it claims there are zero unused deps (this is assumed to be a failure of some sort, and should only happen if there are no deps to begin with).

### CODE REVIEW

**Comment by jakemac53** [comment]
Also apologies for the commit history - I forked the original cl for this work and merging in from master was a lot easier than a rebase.

File: build_modules/lib/src/kernel_builder.dart [hunk 1]
```diff
@@ -131,16 +136,19 @@ Future<void> _createKernel(
...
+    @required bool useIncrementalCompiler,
+    @required bool trackUnusedInputs}) async {
   var request = WorkRequest();
   var scratchSpace = await buildStep.fetchResource(scratchSpaceResource);
   var outputId = module.primarySource.changeExtension(outputExtension);
   var outputFile = scratchSpace.fileFor(outputId);
 
   File packagesFile;
+  File usedInputsFile;
+  Map<String, AssetId> kernelInputPathToId;
```
Comment (natebosch) [inline build_modules/lib/src/kernel_builder.dart]: can we add a doc comment?

File: build_modules/lib/src/kernel_builder.dart [hunk 2]
```diff
@@ -187,11 +204,57 @@ Future<void> _createKernel(
...
+
+    // Note that we only want to do this on success, we can't trust the unused
+    // inputs if there is a failure.
+    if (usedInputsFile != null) {
+      await reportUnusedKernelInputs(
+          usedInputsFile, kernelDeps, kernelInputPathToId, buildStep);
+    }
   } finally {
     await packagesFile.parent.delete(recursive: true);
+    await usedInputsFile.parent.delete(recursive: true);
```
Comment (natebosch) [inline build_modules/lib/src/kernel_builder.dart]: what if this is null?

Comment (jakemac53) [reply build_modules/lib/src/kernel_builder.dart]: it throws! :D

File: build_modules/lib/src/kernel_builder.dart (hunk 1 above)
Comment (jakemac53) [reply build_modules/lib/src/kernel_builder.dart]: I added some doc comments to these both and moved them into their own block (same for the ddc builder)

**Pull request merged**